src/plots/cartesian/set_convert.js

```diff
@@ -371,49 +381,99 @@ module.exports = function setConvert(ax, fullLayout) {
+                            } else {
+                                arrays.push(trace.z);
                             }
+                            valLetter = 'z';
+                        } else if(axLetter === 'y' && trace.x) {
+                            arrays.push(trace.x);
+                            valLetter = 'x';
+                        } else if(trace.y) {
+                            arrays.push(trace.y);
+                            valLetter = 'y';
+                        } else {
+                            var nullArray = arrayIn[0].map(function() {return null;});
+                            arrays.push(nullArray);
+                        }
+                        var objList = sortLib.matrixToObjectList(arrays, cols);
+
+                        Array.prototype.push.apply(fullObjectList, objList);
+
+                        // convert the trace data from list to object and sort (backwards, stable sort)
+                        var sortedObjectList = sortLib.sortObjectList(cols, objList);
+                        var matrix = sortLib.objectListToList(sortedObjectList);
+                        var sortedMatrix = sortLib.sortedMatrix(matrix);
+
+                        axLabels = sortedMatrix[0].slice();
+                        var axVals = sortedMatrix[1];
+
+                        if(valLetter === 'z' & axLetter === 'x') {
+                            axVals = sortLib.transpose(axVals);
+                        }
+
+                        if(trace.type === 'ohlc' | trace.type === 'candlestick') {
+                            var sortedValsTransform = sortLib.transpose(axVals);
+                            gd._fullData[i].open = sortedValsTransform[0];
+                            gd._fullData[i].high = sortedValsTransform[1];
+                            gd._fullData[i].low = sortedValsTransform[2];
+                            gd._fullData[i].close = sortedValsTransform[3];
+                        }
+                        // Could/should set sorted y axis values for each trace as the sorted values are already available.
+                        // Need write access to gd._fullData, bad? Should probably be done right at newPlot, or on setting gd._fullData
+
+                        var transposedAxLabels = sortLib.transpose(axLabels);
+                        if(gd._fullData) {
+                            gd._fullData[i][axLetter] = transposedAxLabels;
+                        }
+                        if(valLetter) {
+                            gd._fullData[i][valLetter] = axVals;
                         }
                     }
                 }
             }
 
-            list.sort(function(a, b) {
-                var ind0 = seen[0][1];
-                var d = ind0[a[0]] - ind0[b[0]];
-                if(d) return d;
+            if(axLabels.length) {
+                ax.levelNr = axLabels[0].length;
+                ax.levels = axLabels[0].map(function(_, idx) {return idx;});
```

Comment:
`ax` is part of `gd._fullLayout` and it's therefore a kind of "cleaned" and "completed" version of `gd.layout`, the `layout` portion of the figure. With that in mind, it's OK to add things to `ax` that don't exist in `layout.xaxis` etc, but only if you prefix them with an underscore.